dm/syncer/validator_checkpoint.go

```diff
@@ -335,34 +305,106 @@ func (c *validatorPersistHelper) persist(tctx *tcontext.Context, loc binlog.Loca
+	count := int(c.validator.getAllPendingRowCount()) + 1
+	queries := make([]string, 0, count)
+	args := make([][]interface{}, 0, count)
+
+	// delete pending row of previous failed persist
+	queries = append(queries, `DELETE FROM `+c.pendingChangeTableName+` WHERE source = ? and revision = ?`)
+	args = append(args, []interface{}{c.cfg.SourceID, rev})
+	// insert pending row changes with revision=rev
+	for _, worker := range c.validator.getWorkers() {
+		for _, tblChange := range worker.getPendingChangesMap() {
+			for key, j := range tblChange.jobs {
+				row := j.row
+				rowForPersist := rowChangeDataForPersist{
+					Key:             key,
+					Tp:              j.Tp,
+					Size:            j.size,
+					Data:            row.RowValues(),
+					FirstValidateTS: j.FirstValidateTS,
+					FailedCnt:       j.FailedCnt,
+				}
+				rowJSON, err := json.Marshal(&rowForPersist)
+				if err != nil {
+					return err
+				}
+				query := `INSERT INTO ` + c.pendingChangeTableName + `
+						(source, schema_name, table_name, row_pk, data, revision) VALUES (?, ?, ?, ?, ?, ?)`
+				queries = append(queries, query)
+				sourceTable := row.GetSourceTable()
+				args = append(args, []interface{}{
+					c.cfg.SourceID,
+					sourceTable.Schema,
+					sourceTable.Table,
+					key,
+					string(rowJSON),
+					rev,
+				})
+			}
+		}
+	}
+	return c.execQueriesWithRetry(tctx, queries, args)
+}
+
+func (c *validatorPersistHelper) persist(tctx *tcontext.Context, loc binlog.Location) error {
 	newCtx, cancelFunc := tctx.WithTimeout(validationDBTimeout)
 	defer cancelFunc()
-	failpoint.Inject("ValidatorCheckPointSkipExecuteSQL", func(val failpoint.Value) {
-		str := val.(string)
-		if str != "" {
-			err = errors.New(str)
-		}
-		failpoint.Goto("afterExecuteSQL")
-	})
-	// baseconn retries too much and with a lot of metric stuff, so use a simpler one.
-	err = c.db.DoTxWithRetry(newCtx, queries, args, c.retryer)
-	failpoint.Label("afterExecuteSQL")
-	if err != nil {
+	// we run sql one by one to avoid potential "transaction too large" error since pending data may quite large.
+	// we use "upsert" to save table status and error row changes, if error happens when persist checkpoint and
+	// pending data, those data maybe inconsistent with each other. but it doesn't matter in validation case,
+	// since the status of the table will be in correct stage after resume and reach the location again.
+	if err := c.persistTableStatusAndErrors(newCtx); err != nil {
 		return err
 	}
 
-	return nil
-}
+	nextRevision := c.revision + 1
 
-func (c *validatorPersistHelper) incrRevision() {
+	// we use "insert" to save pending data to speed up(see https://asktug.com/t/topic/33147) since
+	// the number of pending rows maybe large.
+	// we run sql one by one to avoid potential "transaction too large" error since pending data may quite large.
+	// And use revision field to correspond between checkpoint table and pending row table
```

Comment:
```suggestion
	// And use revision to synchronize versions of checkpoint table and pending row table
```
A bit obscure using `correspond between` (haven't seen such a phrase). I tried paraphrasing it and please let me know if it isn't what you meant.

Reply:
should be link or associate